File: July2021/GrayCode.java
class Solution {

  // TC : O(n2)
  public List<Integer> grayCode(int n) {
    List<Integer> list = new ArrayList<>();
    list.add(0);

    if(n == 0) return list;
    list.add(1);

    int curr = 1;
    for(int i = 2;i<=n;i++){
      curr *= 2;
      for(int j =  list.size()-1;j>=0;j--){
        list.add(curr+list.get(j));
      }
    }
    return list;
  }
}

File: July2021/C++/gray_code.cpp
class Solution {
public:
    void bitflip(bitset<16>& val,vector<int>& res,int k)
    {
        if(k==0)
        {
            // cout<<val<<endl;
            res.push_back(val.to_ulong());
        }
        else
        {
            bitflip(val,res,k-1);
            val.flip(k-1);
            bitflip(val,res,k-1);
        }
    }
    vector<int> grayCode(int n) {
        bitset<16> val;
        vector<int> res;
        bitflip(val,res,n);
        return res;
    }
};